### ISSUE TITLE
Bombard Bug in any map on version 1.9.0.0.5855

### ISSUE THREAD

**Issue by NGMC0001** [triplea-game/triplea#2133]
Games freeze before you receive the window with the option to choose whether or not or which territory you wish to have your ships bombard every time.  Here is the error:
```
triplea.engine.version.bin:1.9
Loading map: new_world_order, from: C:\Users\steve\triplea\downloadedMaps\new_world_order-master.zip
Loading resources from the following paths: [C:\Users\steve\triplea\downloadedMaps\new_world_order-master.zip, C:\Program Files\TripleA\assets]
Loading map: new_world_order, from: C:\Users\steve\triplea\downloadedMaps\new_world_order-master.zip
Loading resources from the following paths: [C:\Users\steve\triplea\downloadedMaps\new_world_order-master.zip, C:\Program Files\TripleA\assets]
java.lang.ArrayIndexOutOfBoundsException: 1
	at games.strategy.engine.message.RemoteMethodCall.classesToString(RemoteMethodCall.java:128)
	at games.strategy.engine.message.RemoteMethodCall.resolve(RemoteMethodCall.java:180)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeSingle(EndPoint.java:153)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeMultiple(EndPoint.java:146)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeLocal(EndPoint.java:132)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.lambda$messageReceived$2(UnifiedMessenger.java:299)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)

```

**Comment by ssoloff** [comment]
Reproduction notes:

* Network game host: 1.9.0.0.3635.
* Network game client: 1.9.0.0.5876.
* Client initiates amphibious attack with naval bombardment during combat movement phase.
* Exception reported by OP occurs at start of combat phase (immediately after finishing production phase).

**Comment by ssoloff** [comment]
It looks like an RMI incompatibility may have been introduced in 385ec5a.

When the exception occurs, the 3635 host is attempting to invoke the `ITripleAPlayer#selectShoreBombard(Territory)` method.  As usual, RMI methods are invoked by ordinal and not by name.

In 385ec5a, a new method was added to `ITripleAPlayer`: `selectProducerTerritoryForUnits(Collection<Unit>, Territory)`.  Based on the RMI sorting rules, this method is ordered before `selectShoreBombard()`, and `selectProducerTerritoryForUnits()` now has the ordinal that `selectShoreBombard()` had in the previous release.  Thus, the 5876 client attempts to invoke the wrong method, and as a result of `selectProducerTerritoryForUnits()` requiring two arguments as opposed to one for `selectShoreBombard()`, an `ArrayIndexOutOfBoundsException` is thrown.

@ron-murhammer Can you please take a look when you have time?  It's not clear to me how to fix this without breaking the dependent code in `AbstractPlaceDelegate#placeUnits()`.

**Comment by ron-murhammer** [comment]
@ssoloff I wasn't aware that was how RMI invoked methods. Could we just rename the method?

**Comment by ssoloff** [comment]
> I wasn't aware that was how RMI invoked methods.

Yeah, it's an "optimization." :smile:  From `RemoteMethodCall`:

```java
// to save space, we dont serialize method name/types
// instead we just serialize a number which can be transalted into
// the correct method.
private int m_methodNumber;
```

> Could we just rename the method?

Possibly.  That was the approach @RoiEXLab took in [#1942](https://github.com/triplea-game/triplea/pull/1942#issuecomment-311488543), and it fixed the issue when an old client connected to a new lobby.  However, we then discovered a problem in the other direction when a new client connected to an old lobby, which was fixed in #2114 by removing the renamed methods from the remote interface to avoid crashing the old lobby.

My point being that if you rename the method with a `zz` prefix or something, you'll need to ensure that a new host doesn't attempt to invoke the new method when connected to a legacy client during a network game, which would then cause a "method not found" exception on the legacy client.

**Comment by ron-murhammer** [comment]
@ssoloff Good point. I think the only option for now is to revert/comment it out to fix compatibility and then re-apply it later on.

**Comment by RoiEXLab** [comment]
We should add the removal of this index to the TODO list of the next incompatible release

**Comment by NGMC0001** [comment]
I appreciate all taking the time to look into the issue.  Might you perhaps
know a timeline for fixing it?

Thank you,
Todd  :)

**Comment by ssoloff** [comment]
@NGMC0001 Ideally, reverting the incompatible change and deploying a new release could be done relatively quickly.  However, there may be other changes made after that one that now depend on it, which could cause a cascade of reversions.  @ron-murhammer should be able to provide a more accurate estimate once he's had time to look into it.

In the meantime, you should be able to work around the problem by running the previous release ([1.9.0.0.3635](https://github.com/triplea-game/triplea/releases/tag/1.9.0.0.3635)), which seems to be the version the other player (or bot) you're playing against is running.  As long as you don't need a particular feature of the latest 5855 release, you should be good.

**Comment by DanVanAtta** [comment]
In response to this I updated the two releases after 1.9.0.0.3635 to be 'prereleases'.

 @NGMC0001 you can now install 1.9.0.0.3635 (which does not have this problem) from the website: http://www.triplea-game.org/download/

**Comment by NGMC0001** [comment]
Thank you. I appreciate the help!

On Jul 26, 2017 3:11 PM, "Dan Van Atta" <notifications@github.com> wrote:

In response to this I updated the two releases after 1.9.0.0.3635 to be
'prereleases'.

@NGMC0001 <https://github.com/ngmc0001> you can now install 1.9.0.0.3635
(which does not have this problem) from the website:
http://www.triplea-game.org/download/

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/triplea-game/triplea/issues/2133#issuecomment-318168337>,
or mute the thread

**Comment by ron-murhammer** [comment]
@ssoloff Feel free to take a shot at fixing this. I'm pretty limited in time at the moment.

**Comment by ssoloff** [comment]
@ron-murhammer NP, will try to dig into it over the next couple of days.

**Comment by ssoloff** [comment]
Just confirming that simply renaming the new method does not work.  As expected, when a new server hosts a network game, it will attempt to invoke the new method on the remote client.  If the remote client is an old version, it will also throw an `ArrayIndexOutOfBoundsException` due to the new method index exceeding the `ITripleAPlayer` method array length.

I'm trying to see if there's a way for one node to determine the engine version of another node, so that it could be temporarily used to bypass the problematic call.  I've been unsuccessful so far, but I'm not really familiar with the RMI code, so I have a lot to go through. 😞

**Comment by NGMC0001** [comment]
Hello,
I've deleted everything TripleA related from my computer and reloaded
version .3635, and the problem persists.  However, it only exists when I
attempt to play online.  I loaded several maps both online and locally, and
the bombard function froze the game every time online and worked every time
locally.  I am not sure why that difference would occur.  Thought it might
help you find the problem (or make it more difficult).  😂  Again, I
appreciate the effort to fix this issue.
Thank you,
Todd

**Comment by ssoloff** [comment]
@NGMC0001 Sorry for not explicitly stating this earlier, but, as you observed, the problem (well, a similar problem with the same root cause) also occurs if the network host is 1.9.0.0.5855 and the network client is 1.9.0.0.3635.

When you first opened this issue, I wasn't aware that the new engine code had been deployed as lobby bots, but it apparently has (see the discussion in #2147).  If I had known, I wouldn't have suggested that downgrading to 3635 would fix your problem. 😞  Unfortunately, I don't have any visibility into the status of the bot rollback in the lobby.  Maybe @prastle or @DanVanAtta can comment when the old engine has been restored.

> I am not sure why that difference would occur.

This may totally bore you, but because you asked... :smile:  The reason why the problem only occurs online is because two different versions of TripleA are talking to each other over the network.  There are specific rules in place for how these two nodes must send data back and forth.  Unfortunately, the newer version is breaking one of those rules, which is causing the errors you've seen.

When you play the game locally, there is no network communication between two different versions of TripleA--all data is moved around within a single version of the application.  Because the bug is only in the network code, and a local game never executes the network code in question, the bug is never hit, and you can bombard to your heart's content. :smile:

> Again, I appreciate the effort to fix this issue.

Thanks for your patience here.  A fix for the bug has been submitted in #2148.  Unfortunately, that's not going to help with the immediate problem until the lobby bots are downgraded to the previous version.  Hopefully that will happen soon (if not already complete).

**Comment by NGMC0001** [comment]
No apologies necessary.  I am just glad you have all taken your valuable
time to look into this and reply in such a swift manner.  I am confident a
fix will be in place soon (if not already... either through Bot changes or
programming changes).  Thanks!

**Comment by DanVanAtta** [comment]
bots are all back to 3635 now

**Comment by NGMC0001** [comment]
Thanks. I'll try it out tonight.

**Comment by ssoloff** [comment]
The fix for this issue has been merged and is available in build 1.9.0.0.5947 or later.

Since we've downgraded the bots back to 1.9.0.0.3635, there's not much point in asking users to test the fix from #2148.  However, if anyone wants to independently verify the fix with 5947, that would be great!

Otherwise, I'll leave this issue open until @NGMC0001 verifies the bot downgrade also works.

**Comment by NGMC0001** [comment]
Hello,

I'm happy to report the bot downgrade worked.  I was able to enjoy a gaming
session last night.  I appreciate all the attention to this issue.  The
response was swift and successful!

Thank you,
Todd
